File: web/src/types/kpi.ts
export interface KPIMetricValue {
  metric_name: string;
  value: number;
  unit: string;
  agent_role: string | null;
  recorded_at: string | null;
}

export interface KPISummary {
  total_tasks: number;
  completed_tasks: number;
  success_rate: number;
  avg_duration_minutes: number;
  total_tokens: number;
  total_cost_rmb: number;
  metrics: KPIMetricValue[];
}

export interface KPITimeSeriesPoint {
  timestamp: string;
  value: number;
}

export interface KPITimeSeries {
  metric_name: string;
  unit: string;
  data: KPITimeSeriesPoint[];
}

// ── ROI Dashboard ──────────────────────────────────────

export interface ROITaskBreakdown {
  task_id: string;
  title: string;
  agent_cost_rmb: number;
  estimated_manual_rmb: number;
  savings_rmb: number;
  agent_duration_minutes: number;
  estimated_manual_hours: number;
}

export interface AgentRoleEfficiency {
  role: string;
  display_name: string;
  total_stages: number;
  total_tokens: number;
  avg_duration_seconds: number;
  total_cost_rmb: number;
}

export interface ROISummary {
  total_tasks_completed: number;
  total_agent_cost_rmb: number;
  total_estimated_manual_rmb: number;
  total_savings_rmb: number;
  roi_ratio: number;
  total_agent_hours: number;
  total_estimated_manual_hours: number;
  time_saved_hours: number;
  benchmark_hours_per_task: number;
  benchmark_hourly_rate: number;
  by_role: AgentRoleEfficiency[];
  recent_tasks: ROITaskBreakdown[];
}

// ── Developer Cockpit ──────────────────────────────────

export interface CockpitTaskItem {
  id: string;
  title: string;
  status: string;
  project_name: string | null;
  template_name: string | null;
  created_at: string;
  completed_at: string | null;
  current_stage: string | null;
  error_message: string | null;
  total_tokens: number;
  total_cost_rmb: number;
}

export interface CockpitData {
  pending_gates_count: number;
  running_tasks_count: number;
  failed_tasks_today: number;
  completed_tasks_today: number;
  pending_gates: Array<{
    id: string;
    gate_type: string;
    task_id: string;
    agent_role: string;
    content: Record<string, string> | null;
    status: 'pending' | 'approved' | 'rejected';
    reviewer: string | null;
    review_comment: string | null;
    created_at: string;
    reviewed_at: string | null;
  }>;
  running_tasks: CockpitTaskItem[];
  failed_tasks: CockpitTaskItem[];
  recent_completed: CockpitTaskItem[];
}

File: platform/app/schemas/kpi.py
from __future__ import annotations

from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel


class KPIMetricValue(BaseModel):
    metric_name: str
    value: float
    unit: str
    agent_role: Optional[str] = None
    recorded_at: Optional[datetime] = None


class KPISummaryResponse(BaseModel):
    total_tasks: int = 0
    completed_tasks: int = 0
    success_rate: float = 0.0
    avg_duration_minutes: float = 0.0
    total_tokens: int = 0
    total_cost_rmb: float = 0.0
    metrics: List[KPIMetricValue] = []


class KPITimeSeriesPoint(BaseModel):
    timestamp: datetime
    value: float


class KPITimeSeriesResponse(BaseModel):
    metric_name: str
    unit: str
    data: List[KPITimeSeriesPoint]


class KPIReportResponse(BaseModel):
    generated_at: datetime
    period: str
    summary: KPISummaryResponse
    by_agent: Dict[str, KPISummaryResponse] = {}


# ── ROI Dashboard ──────────────────────────────────────────


class ROITaskBreakdown(BaseModel):
    task_id: str
    title: str
    agent_cost_rmb: float
    estimated_manual_rmb: float
    savings_rmb: float
    agent_duration_minutes: float
    estimated_manual_hours: float


class AgentRoleEfficiency(BaseModel):
    role: str
    display_name: str
    total_stages: int
    total_tokens: int
    avg_duration_seconds: float
    total_cost_rmb: float


class ROISummaryResponse(BaseModel):
    total_tasks_completed: int
    total_agent_cost_rmb: float
    total_estimated_manual_rmb: float
    total_savings_rmb: float
    roi_ratio: float
    total_agent_hours: float
    total_estimated_manual_hours: float
    time_saved_hours: float

    benchmark_hours_per_task: float
    benchmark_hourly_rate: float

    by_role: List[AgentRoleEfficiency]
    recent_tasks: List[ROITaskBreakdown]


# ── Developer Cockpit ──────────────────────────────────────


class CockpitTaskItem(BaseModel):
    id: str
    title: str
    status: str
    project_name: Optional[str] = None
    template_name: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    current_stage: Optional[str] = None
    error_message: Optional[str] = None
    total_tokens: int
    total_cost_rmb: float

    model_config = {"from_attributes": True}


class CockpitResponse(BaseModel):
    pending_gates_count: int
    running_tasks_count: int
    failed_tasks_today: int
    completed_tasks_today: int

    pending_gates: List["GateDetailResponse"]
    running_tasks: List[CockpitTaskItem]
    failed_tasks: List[CockpitTaskItem]
    recent_completed: List[CockpitTaskItem]


# Avoid circular import – import at module level is fine because
# gate.py has no reverse dependency on kpi.py.
from app.schemas.gate import GateDetailResponse  # noqa: E402

CockpitResponse.model_rebuild()
